packages/go_router/lib/src/misc/stateful_navigation_shell.dart

```diff
@@ -0,0 +1,422 @@
+        _routeState.currentBranchState;
+    _navigatorCache.clear();
+    _setupInitialStatefulShellRouteState();
+    GoRouter.of(context).go(navigatorState.branch.defaultLocation);
+  }
+
+  StatefulShellBranchState _copyStatefulShellBranchState(
+      StatefulShellBranchState branchState,
+      {BranchNavigatorBuilder? navigatorBuilder,
+      UnmodifiableRouteMatchList? matchList,
+      bool? loaded}) {
+    bool dirty = branchState.matchList != matchList;
+    if (navigatorBuilder != null) {
+      final Navigator? existingNav = _navigatorForBranch(branchState.branch);
+      if (existingNav == null || dirty) {
+        dirty = true;
+        _setNavigatorForBranch(branchState.branch, navigatorBuilder());
+      }
+    }
+
+    final _BranchNavigatorProxy branchWidget =
+        branchState.child as _BranchNavigatorProxy;
+    final bool isLoaded =
+        loaded ?? _navigatorForBranch(branchState.branch) != null;
+    dirty = dirty || isLoaded != branchWidget.loaded;
+
+    if (dirty) {
+      return branchState.copy(
+        child: branchWidget.copy(loaded: isLoaded),
+        matchList: matchList,
+      );
+    } else {
+      return branchState;
+    }
+  }
+
+  StatefulShellBranchState _createStatefulShellBranchState(
+      StatefulShellBranch branch,
+      {BranchNavigatorBuilder? navigatorBuilder,
+      UnmodifiableRouteMatchList? matchList}) {
+    if (navigatorBuilder != null) {
+      _setNavigatorForBranch(branch, navigatorBuilder());
+    }
+    return StatefulShellBranchState(
+      branch: branch,
+      child: _BranchNavigatorProxy(
+        branch: branch,
+        navigatorForBranch: _navigatorForBranch,
+      ),
+      matchList: matchList,
+    );
+  }
+
+  void _setupInitialStatefulShellRouteState() {
+    final List<StatefulShellBranchState> states = widget.branches
+        .map((StatefulShellBranch e) => _createStatefulShellBranchState(e))
+        .toList();
+
+    _routeState = StatefulShellRouteState(
+      route: widget.shellRoute,
+      branchStates: states,
+      currentIndex: 0,
+      switchActiveBranch: _switchActiveBranch,
+      resetState: _resetState,
+    );
+  }
+
+  @override
+  void initState() {
+    super.initState();
+    _setupInitialStatefulShellRouteState();
+  }
+
+  @override
+  void didUpdateWidget(covariant StatefulNavigationShell oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    _updateRouteStateFromWidget();
+  }
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    _updateRouteStateFromWidget();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return InheritedStatefulNavigationShell(
+      routeState: _routeState,
+      child: Builder(builder: (BuildContext context) {
+        // This Builder Widget is mainly used to make it possible to access the
+        // StatefulShellRouteState via the BuildContext in the ShellRouteBuilder
+        final ShellRouteBuilder shellRouteBuilder = widget.shellRoute.builder!;
```

Comment:
Using only builder. Param `pageBuilder` never used.

Reply:
Yes, the `StatefulNavigationShell` only uses the `builder` in its build method, since it is only concerned with building the actual navigation shell at that point. The `pageBuilder` is used when building the `Page`s for the Navigator in `RouteBuilder`, which is done the same way as for other routes: 

https://github.com/flutter/packages/blob/e9d40eee9ba0c7d811a935ded60a6a0c3bc09107/packages/go_router/lib/src/builder.dart#L235-L237